client/session.go

```diff
@@ -574,99 +651,67 @@ func (s *session) Write(namespace, id string, t time.Time, value float64, unit x
 	}
 
-	majority = atomic.LoadInt32(&s.majority)
+	c := s.writeStatePool.Get().(*writeState)
+	c.incref()
 
-	w := s.writeOpPool.Get()
-	w.namespace = ts.StringID(namespace)
-	w.request.ID = tsID.Data()
-	w.request.Datapoint.Value = value
-	w.request.Datapoint.Timestamp = timestamp
-	w.request.Datapoint.TimestampType = timeType
-	w.request.Datapoint.Annotation = annotation
-	w.completionFn = func(result interface{}, err error) {
-		var snapshotSuccess int32
-		if err != nil {
-			atomic.AddInt32(&resultErrs, 1)
-			resultErrLock.Lock()
-			errors = append(errors, err)
-			resultErrLock.Unlock()
-		} else {
-			snapshotSuccess = atomic.AddInt32(&success, 1)
-		}
-		remaining := atomic.AddInt32(&pending, -1)
-		doneAll := remaining == 0
-		switch s.writeLevel {
-		case topology.ConsistencyLevelOne:
-			complete := snapshotSuccess > 0 || doneAll
-			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		case topology.ConsistencyLevelMajority:
-			complete := snapshotSuccess >= majority || doneAll
-			if complete && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		case topology.ConsistencyLevelAll:
-			if doneAll && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		}
-		// Finalize resources that can be returned
-		if doneAll {
-			// Return write to pool
-			s.writeOpPool.Put(w)
-		}
-	}
+	c.op, c.majority = s.writeOpPool.Get(), majority
 
-	routeErr := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
+	c.op.namespace = ts.StringID(namespace)
+	c.op.request.ID = tsID.Data()
+	c.op.request.Datapoint.Value = value
+	c.op.request.Datapoint.Timestamp = timestamp
+	c.op.request.Datapoint.TimestampType = timeType
+	c.op.request.Datapoint.Annotation = annotation
+	c.op.completionFn = c.completionFn
+
+	if err := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
 		// First count all the pending write requests to ensure
 		// we count the amount we're going to be waiting for
 		// before we enqueue the completion fns that rely on having
 		// an accurate number of the pending requests when they execute
-		pending++
-	})
-	if routeErr == nil {
-		// Now enqueue the write requests
-		routeErr = s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
-			enqueued++
-			if err := s.queues[idx].Enqueue(w); err != nil && enqueueErr == nil {
-				// NB(r): if this ever happens we have a code bug, once we
-				// are in the read lock the current queues we are using should
-				// never be closed
-				enqueueErr = err
-			}
-		})
+		c.pending++
+		c.queues = append(c.queues, s.queues[idx])
+	}); err != nil {
+		c.decref()
+		s.RUnlock()
+		return err
 	}
-	s.RUnlock()
 
-	if routeErr != nil {
-		return routeErr
-	}
+	c.Lock()
 
-	if enqueueErr != nil {
-		s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", enqueueErr)
-		return enqueueErr
+	for i := range c.queues {
+		if err := c.queues[i].Enqueue(c.op); err != nil {
```

Comment:
Yeah I think we had it there with the intention that if we had a bug somewhere that caused the queues to be closed before/during enqueuing, we'd be able to see it in the logs, which would make it easier to debug issues. Don't think that's necessary now that we've done extensive testing on the write path so I'm fine with taking that part out.